keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,345 @@
+- For memory, total amount of sefety zone memory(value of safety zone memory * number of NUMA nodes) cannot be greater than total amount of reserved memory(kube-reserved, system-reserved and eviction-threshold) by Node Allocatable feature.
+
+- For hugepages, at the moment(v1.17), Node Allocateble features does not support reserving hugepages, so that Memory Manager will not have safety zone for hugepages. Once Node Allocatable feature will support hugepages reservation([#83541](https://github.com/kubernetes/kubernetes/pull/83541)), safety zone for hugapages will works in the same manner of memory.
+
+For example, if Node Allocatable feature reserved total 10Gbi memory on node which is two sockets server machine, safety zone memory cannot be greater than 5Gbi.
+
+```
+                  +-----------------+-------------------------+
+NUMA node0(40Gi)  | Safety Zone(5Gi)|    Allocatable(35Gi)    |
+                  +-----------------+-------------------------+
+                  +-----------------+-------------------------+
+NUMA node1(40Gi)  | Safety Zone(5Gi)|    Allocatable(35Gi)    |
+                  +-----------------+-------------------------+
+```
+_Figure: The example of safety zone._
+
+
+#### The Concept of Reserved Zone
+
+Memory Manager has the concept of reserved zone to calculate NUMA node affinity precisely. The reserved zone is literally reserved for the containers which has Guaranteed QoS. Container doesn't always consume memory as much as they requested. For example container may request 5Gbi memory, but it can consumes only 3Gbi at runtime. Memory Manager treats requested memory of Guaranteed QoS containers as reserved memory. Memory Manager guarantee Guaranteed QoS containers usage of memory and hugepages on single NUMA node as they requested by using [CPUSET enforcement](#cpuset-enforcement) and [OOM Score Enforcement](#oom-score-enforcement). 
+
+```
+           +---------------+------------+--------------+
+           |               |       Allocatable         |
+NUMA node  +  Safety Zone  +------------+--------------+
+           |               |  Reserved  |     free     |
+           +---------------+------------+--------------+
+```
+_Figure: The memory map of NUMA node._
+
+NOTE:
+- Reserved zone cannot be greater than Allocatable zone.
+- Typical Linux processes and containers which have BestEffort or Burstable QoS may consume memory on a NUMA node where Guaranteed QoS containers also consumes memory. It is also possible that Guaranteed QoS containers requested lots of memory but consumes just a little bit of them and the others occupy reserved zone when only few free memory is available on the NUMA node. At this situation, OOM may happen by Guaranteed QoS Container if the container try to consume more memory. Then others will be killed by OOM killer to make free memory. Because, unlike Guaranteed QoS containers, the others has high OOM socre(which has low priority at OOM system).
```

Comment:
When the OOM-killer runs, is there any performance hit to the low-latency DPDK app running in the Guaranteed QoS pod?  If so, that seems like a potential problem.  Seems like it'd be better to prevent the OOM-killer from running in the first place.  That might require running each pod on a single NUMA node to ensure that the available memory per NUMA node is known, and enforce hard limits on how much memory can be allocated even by Burstable or BestEffort pods.

Reply:
I will comment on it after the meeting.

Reply:
In my mind, I don't think OOM-killer afftects the performance.
Kernel & containers(in G-QoS Pod with exclusive CPU usage) will consume different CPUs.
By node allocatable feature and `--reserved-cpus` flag, we can assign CPUs for system(includes kernel) & kubernetes core things explictly.
Just like `--reserved-cpus=0-3` in '--system-reserved', it means admistrator assignes CPU 0, 1, 2, 3 for the system.

Reply:
But I also agree that it can be potential problem.
I will perfome a kind of feasibility test for Memory Manager in Feb.
I think, triggering OOM while DPDK runing then measuring the performance of DPDK, it can be a good test scenario.
I will add it to test scenario.

Reply:
Some years back when I did testing, when the OOM-killer ran it caused latency spikes on the whole system.  I don't know if that's still the case...has anyone tested what cyclictest looks like on one CPU when the OOM killer runs on another CPU?

Reply:
@cbf123
My team performed a test but we couldnt find latency spikes when global oom event is triggered.
I tested with pkgen & testpmd.
As i said above, i reserved cpus(two) for system-reserved.